packbackbooks/lti-1-3-php-library
# PODA-7: Update the LTI 1.3 Library to cache auth tokens

## Summary of Changes

Implements Guzzle and caches auth tokens. 

## Testing

- [x] I have run `composer test`
- [x] I have run `composer lint-fix` 
- In `backend/app-api/composer.json` I set  `"packbackbooks/lti-1p3-tool": "dev-cache-auth-tokens",` to point the BE to a specific library branch. 
  - then run `composer install` and `make pull` 
- Then seeded GBS Data `php artisan db:seed --class=GradebookSyncDataSeeder`
- launch an assignment. 
  - Log into canvas using a user from the GBS Dataseeder and open up an assignment, make sure the user is associated to a packback user.
- Exec'd into docker container (`docker-compose exec app-api-php-fpm bash`) and opned a tinker session (`php artisan tinker`)
- Then queued grades and synced 

```
foreach(Grade::where('community_id', 'c0000012-0000-8000-0000-000000000000')->get() as $grade) if ($grade->canBeSynced) SyncGradesForLtiKey::queueGradeToSync($grade);

foreach (KeyDetails::get() as $key) Queue::push(new SyncGradesForLtiKey($key->id));
```
Was able to sync grades without an error about the access token being expired or a bad request from the guzzle client. 

I did run into this error a lot when syncing grades

```
{\"errors\":{\"type\":\"unprocessable_entity\",\"message\":\"User not found in course or is not a student\"}}
today at 12:49:57 PM ","lti_key":7}
```

[It's a normal error ](https://packbackco.slack.com/archives/C01L8QXBJF4/p1623968265229000) seems to happen if a user is linked to packback correctly. 

## PR Dependencies

- [ ] https://github.com/packbackbooks/questions/pull/4455
